ethabi/src/param_type/reader.rs

```diff
@@ -6,38 +6,76 @@ pub struct Reader;
 impl Reader {
 	/// Converts string to param type.
 	pub fn read(name: &str) -> Result<ParamType, Error> {
-		// check if it is a fixed or dynamic array.
-		if let Some(']') = name.chars().last() {
-			// take number part
-			let num: String = name.chars()
-				.rev()
-				.skip(1)
-				.take_while(|c| *c != '[')
-				.collect::<String>()
-				.chars()
-				.rev()
-				.collect();
+		match name.chars().last() {
+            // check if it is a struct
+            Some(']') if name.starts_with('[') => {
+                let mut subtypes = Vec::new();
+                let mut nested = 0isize;
+                let mut last_item = 1;
 
-			let count = name.chars().count();
-			if num.is_empty() {
-				// we already know it's a dynamic array!
-				let subtype = try!(Reader::read(&name[..count - 2]));
-				return Ok(ParamType::Array(Box::new(subtype)));
-			} else {
-				// it's a fixed array.
-				let len = try!(usize::from_str_radix(&num, 10));
-				let subtype = try!(Reader::read(&name[..count - num.len() - 2]));
-				return Ok(ParamType::FixedArray(Box::new(subtype), len));
-			}
-		}
+                for (pos, c) in name.chars().enumerate() {
+                    match c {
+                        '[' => {
+                            nested += 1;
+                        }
+                        ']' => {
+                            nested -= 1;
+                            if nested < 0 {
+                                return Err(ErrorKind::InvalidName(name.to_owned()).into());
+                            } else if nested == 0 {
+                                let sub = &name[last_item..pos];
+                                let subtype = Reader::read(sub)?;
+                                subtypes.push(Box::new(subtype));
+                                last_item = pos + 1;
+                            }
+                        }
+                        ',' if nested == 1 => {
+                            let sub = &name[last_item..pos];
+                            let subtype = Reader::read(sub)?;
+                            subtypes.push(Box::new(subtype));
+                            last_item = pos + 1;
+                        }
+                        _ => (),
+                    }
+                }
+                return Ok(ParamType::Tuple(subtypes));
```

Comment:
Should we not use dedicated parsing with `(` and `)` for tuples rather than piggybacking on the array syntax?